### ISSUE TITLE
Automatic GPU offloading 

### ISSUE THREAD

**Issue by jon-chuang** [rayon-rs/rayon#798]
It seems to me that rayon's awareness of iterators, types and mutability ought to make automatic GPU offloading a feasible project. 

An idea would be to use compiler plugins/proc-macros that precompile some of rayon's data structures to parse the `par_iter` (just via brute force e.g. pattern matching in ast) as a kernel and data movement, and automatically generate the wgpu-rs code necessary, and by isolating the kernel, and precompiling it into LLVM-IR, one can do Spir-V code generation at compile time.

One can also look into composing kernels the same way iterators can be composed. Currently, is rayon able to compose closures such as `.map(|x| x*2).map(|x| if x % 32 == 0 { x } else { x + 1})`?

@calebwin maybe relevant to you

**Comment by Walther** [comment]
This could be super interesting!
I have a toy raytracer project where I would be delighted to play around with this. Sadly, I don't have much knowledge of either rayon's or spir-v's internals, or general purpose GPU compute kernels, to be of much help on the implementation side, without significant help.

**Comment by dlight** [comment]
OpenMP, which is similiar in spirit to Rayon, is able to offload portions of C and C++ programs to GPUs. Maybe Rayon could leverage the existing OpenMP infrastructure instead of building one from scratch. But being able to compile Rust to SPIR-V, nvptx or amdgpu is also needed.

There's also sycl that leverages OpenCL to provide a high-level way to build heterogeneous C++ programs that runs on both CPU and GPU; could be an interesting model for a Rust library to based on.

**Comment by Walther** [comment]
Possibly relevant comment in this issue https://github.com/rayon-rs/rayon/issues/778 
> There's an unstable target for nvptx64-nvidia-cuda, but it doesn't seem like anyone is working on it.
> Here's the tracking bug: rust-lang/rust#38789

**Comment by Walther** [comment]
Related and probably of interest: https://github.com/EmbarkStudios/rust-gpu

**Comment by rcarson3** [comment]
So as someone who's spent their fair share of time working with HPC applications and various abstraction layers/libraries to get things onto the GPU within C++ and actually run with a nice speed boost, it doesn't appear to me that Rayon is an appropriate crate for such GPU offloading. Another crate would be more suitable that's capable at the very least of fusing small kernels together and targeting various backends (CUDA, HIP, SYCL, and etc). Data management between the device and host code would probably also need to be its own crate as well seeing how that can be quite complex of an issue as well. You could probably look what's being done over in C++ land for some ideas for how to accomplish this ([CARE](https://github.com/LLNL/CARE), [Umpire](https://github.com/LLNL/Umpire), [RAJA](https://github.com/LLNL/RAJA), [Kokkos](https://github.com/kokkos/kokkos), and so many others). Honestly, the field is evolving quite a bit at this time for what these abstraction layers would look like, and we're still figuring out ways to improve them and make them easier to use.

**Comment by dlight** [comment]
But why can't rayon just replicate whatever OpenMP does? Down to using the same LLVM mechanisms that implement OpenMP.

This is already done! You don't need to bring CUDA or other stuff to the picture.

Here's a link: [Can OpenMP be used for GPUs?](https://stackoverflow.com/questions/28962655/can-openmp-be-used-for-gpus)

**Comment by cuviper** [comment]
Rayon is a pure library, with no compiler integration at all.

**Comment by rcarson3** [comment]
So, I'm well aware of the use of OpenMP for target offloading. It's pretty much the only reasonable way to get Fortran code to run on the GPU. You can even get fairly decent speed-ups using OpenMP target-offloading if you do everything right. However, OpenMP is deeply integrated with internals related to the compiler. I mean the C/C++/Fortran versions all require compiler directives (#pragma/!$OMP) for example. I don't know of many people who like monkey around with compiler intrinsic in order to get things to work. 

If the Rayon team wanted to they definitely could go down that route, but I would imagine that would also require them working more with the rust compiler team to get things up and going. Alternatively, another crate could be formed to take up that task as well.

**Comment by dlight** [comment]
@cuviper yes, but perhaps it should hook into the compiler. Other languages do this, why not Rust?

(But in this case, this issue is filed at the wrong repo)

**Comment by dlight** [comment]
But another route would be to write a Rayon-like GPU library with [rust-gpu](https://github.com/EmbarkStudios/rust-gpu), and make it an optional dependency of Rayon.

This sounds more actionable (as in, work in a MVP could be started right now), but currently rust-gpu is very restricted.

**Comment by jon-chuang** [comment]
@dlight I am in agreement, I think rust-gpu would be a very good route and one can even start experimenting immediately on simple examples. 

@rcarson3 I think in this case kernel fusion could possibly be automatic, this depends on how inlining is handled. But probably need to look into how rust-gpu handles functions with spir-v in greater depth.

**Comment by rcarson3** [comment]
@jon-chuang it might make more sense to move this thread over to another repo (such as the [rust-cuda/wg](https://github.com/rust-cuda/wg)) or maybe start a new one up rather than keeping things here in this Rayon thread.

Also, I can't say I've really I'm really familiar with how web-gpu does things, since it's never been on my radar for usage in scientific computing. At least in that field a majority of the work in that field has up until recently been geared towards using CUDA. Although, things might now change with HIP and SYCL being a thing. I probably won't have the time to invest into this new Rust library, but I would strongly suggest looking into what's already been done in the C++ field for these type of things. A lot of time and research has been invested over there, and it wouldn't hurt to use those as a starting point for a Rust library.

**Comment by jon-chuang** [comment]
@rcarson3 sorry, I don't appreciate your bland and uninsightful opinions

**Comment by cuviper** [comment]
@jon-chuang - please keep it civil here. Better to say nothing if you have no constructive response.

**Comment by cuviper** [comment]
> @cuviper thoughts on opening an experimental branch on this repo for that once some work is underway?

Anyone is free to experiment and collaborate on branches of their own fork. Once you have something to show, we can talk about whether that has an integration path back into the main Rayon.

**Comment by jon-chuang** [comment]
@cuviper thanks. 

I'm just expressing my honest opinion. The other party has done nothing but spout irrelevant facts that have little whatsover to do with the topics at hand. His reply is also completely impertinent to the comments I made. Hence, my preference is if he leave the discussion altogether. 

What's the point of being civil if there's no honesty involved?

**Comment by Jasper-Bekkers** [comment]
While everybody in the rust-gpu project is currently on holidays - feel free to reach out to us on our public Discord server (https://discord.gg/dAuKfZS) to see if we can align on this project.

**Comment by jon-chuang** [comment]
Some interesting work:
GPU work stealing (2011): http://www.owlnet.rice.edu/~jmg3/wstgpu.pdf

**Comment by LifeIsStrange** [comment]
For inspiration, the [tornadoVM](https://github.com/beehive-lab/TornadoVM) is likely the state of the art for the JVM and one of its strength is that it is free to choose between offloading on a workload on the gpu or on other accelerators such as AVX.

**Comment by Corallus-Caninus** [comment]
Tensorflow may be interesting for this since kernels can be lazily evaluated into a single graph/scope using something similar to thread_local macro for tensorflow sessions. Tensorflow-rs would need ops::constant to impl into and from for native types (i32 etc) and impls for ops::add ops::mult to rusts std::ops::add etc. If done correctly this may be efficient since every iter would be loaded in the graph already at runtime.

**Comment by eliasboegel** [comment]
I must agree with @rcarson3 here. There has been a monumental amount of funded work going into this topic in the form of several approaches. I think the best starting point there would be to have a look at [Kokkos](https://github.com/kokkos/kokkos), particularly as they essentially have achieved mapping of a single-source kernel to both CPUs and GPUs with a library-only approach. A compiler-based approach (e.g. SYLC) is another option, which for Rust is somewhat easier to deal with than for C++, as in Rust only one single compiler is widely used instead of several different vendor compilers. Kokkos (and the other libraries @rcarson3 mentioned) enjoy the benefits of having had a lot of engineering work go into them, with a lot of lessons learned along the way from many different architectures. Moreover, there is some cooperation between hardware vendors and the developers of the large parallel programming models. I don't see any reason to start any design work from scratch instead of aligning with the existing and mature work in this topic.

**Comment by nielsle** [comment]
Related thought: It could be interesting to have a version of rayon that works with MPI.
